cosmos/web/views.py

```diff
@@ -57,23 +58,126 @@ def workflow(name):
+        first_url = url_for("cosmos.stage", workflow_name=workflow_name, stage_name=stage_name,
+                            page=1, keyword=keyword,
+                            sorting=sorting, order=order) if page != 1 else None
+        prev_url = url_for("cosmos.stage", workflow_name=workflow_name, stage_name=stage_name,
+                           page=page - 1, keyword=keyword,
+                           sorting=sorting, order=order) if page >= 2 else None
+        next_url = url_for("cosmos.stage", workflow_name=workflow_name, stage_name=stage_name,
+                           page=page + 1, keyword=keyword,
+                           sorting=sorting, order=order) if page < max_page else None
+        last_url = url_for("cosmos.stage", workflow_name=workflow_name, stage_name=stage_name,
+                           page=max_page, keyword=keyword,
+                           sorting=sorting, order=order) if page != max_page else None
+
+        # this will change only the url for the column currently used for sorting
+        order_cycle = {None: "asc", "asc": "desc", "desc": None}
+        ordering_for_urls = {colname: order_cycle[order] if good == sorting else "asc"
+                             for colname, good in zip(colnames, names_good)}
+        ordering_urls = {colname: url_for(f"cosmos.stage", workflow_name=workflow_name, stage_name=stage_name,
+                                          page=page, keyword=keyword, sorting=good, order=ordering_for_urls[colname]
+                                          ) if good else None
+                         for colname, good in zip(colnames, names_good)}
+
+        submitted = [t for t in tasks_paginated if t.status == TaskStatus.submitted]
         jm = JobManager(get_submit_args=None, logger=None)
 
         f = attrgetter("drm")
         drm_statuses = {}
         for drm, tasks in it.groupby(sorted(submitted, key=f), f):
             drm_statuses.update(jm.get_drm(drm).drm_statuses(list(tasks)))
 
-        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses)
+        url_handle = url_for('cosmos.handle_stage', old_page=page, old_keyword=keyword, sorting=sorting, order=order,
+                             workflow_name=workflow_name, stage_name=stage_name)
+
+        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses,
+                               tasks_on_page=tasks_paginated, max_page=max_page, colnames=colnames,
+                               ordering_urls=ordering_urls, page=page, url_handle=url_handle,
+                               first_url=first_url, prev_url=prev_url, next_url=next_url, last_url=last_url,
+                               workflow_name=workflow_name, stage_name=stage_name, keyword=keyword)
         # x=filter(lambda t: t.status == TaskStatus.submitted, stage.tasks))
 
+    @bprint.route("/workflow/<workflow_name>/<stage_name>/handle_stage", methods=['GET', 'POST'])
```

Comment:
"handle_stage" is very ambiguous, maybe rename to "query"?

Reply:
or "paginate"? what do you think is clearer

Reply:
its not supe robvious to me why you need a second function who's only purpose is to redirect.  if you think its necessary, can you write a comment why?

Reply:
ok, changed to query

Reply:
I use it to prevent resending the form on page refresh